File: arrays/53. maximum-subarray/maximum_subarray.rb
class Solution
  """Leetcode https://leetcode.com/problems/maximum-subarray problem's solution"""
  def maxSubArray(arr)
    # initailize max_end_hear and max_so_far hear with 0
    max_end_hear = max_so_far = 0

    arr.each do | a |
      max_end_hear += a
      if max_end_hear < 0
        ## element -ve then max_end_hear will be 0
        max_end_hear = 0
      end
      # update max_so_far whe max_end_hear is higer than max_so_far
      if max_end_hear > max_so_far
        max_so_far = max_end_hear
      end
    end
    max_so_far
  end
end

## Solution.new.maxSubArray([-2, -3, 4, -1, -2, 1, 5, -3])

File: arrays/53. maximum-subarray/maximum_subarray.py
import sys


class Solution:
    """Leetcode https://leetcode.com/problems/maximum-subarray problem's solution"""

    def maxSubArray(self, nums: List[int]) -> int:
        """
        Finding the maximum subarray for given input array

        Returns:
            int: Maximum subarray for given input array
        """
        running_sum = 0  # current sum
        max_sum = 0  # maximum sum
        max_number = -sys.maxsize - 1  # max element in the list

        # iterate the input list
        for element in nums:
            # find maximum element in the input list
            if element > max_number:
                max_number = element

            # finding the maximum sub array - Greedy Approach
            running_sum += element
            running_sum = max(running_sum, element)
            max_sum = max(running_sum, max_sum)

        # if max element is negative then return that element
        if max_number < 0:
            return max_number

        return max_sum
